CONTRIBUTING.rst

```diff
@@ -411,6 +411,55 @@ encouraged to use the ``.. versionadded`` sphinx directive. For example:
         """
 
 
+Input Validation
+^^^^^^^^^^^^^^^^
+Validating user input is an essential part of software development to ensure
+algorithms perform as expected. This is especially the case in Python when
+working with multiple array-like object types (e.g. ``Sequence`` and
+``numpy.ndarray``) which may lack support for explicit typing and type
+hints for cases where arrays with a certain shape, dimensionality, and/or
+data type (e.g. ``float``, ``int``) may be required. Therefore it is imperative
+to validate all user input and raise any errors as appropriate.
+
+However, manually writing input validation routines and associated test cases
+can be time consuming and error prone. Therefore, it is recommended to make
+use of existing validation methods in ``pyvista.core.input_validation``.
+
+For example, a typical validation routine may look like:
+
+.. code:: python
+
+    def some_function(points_array, method):
+        """ Apply a method to an Nx3 points array.
+
+        Parameters
+        ----------
+        points_array : array_like
+            Array-like input of points with shape Nx3.
+
+        method : str
+            Method to apply. Must be one of:
+                * ``'method_a'``
+                * ``'method_b'``
+
+        """
+        import pyvista.core.input_validation as valid
+
+        # Validate input array. There is no need to check for
+        # type (e.g. Sequence or np.ndarray) or array shape.
+        # An error is automatically raised for bad input.
+        arr = valid.validate_arrayNx3(points_array)
+
+        # Validate input method. An error is automatically
+        # raised if the method is not valid.
+        possible_methods = ["method_a", "method_b"]
+        check_is_string_in_list(method, possible_methods)
+
+When using the array validation methods, consider setting the input
+type in the docstring to ``array_like`` (if appropriate), as the validation
```

Comment:
is `array_like` defined somewhere, or it's just a convention?

Reply:
It's not formally defined as a type, no. The term `array_like` is just a [NumPy glossary term](https://numpy.org/devdocs/glossary.html#term-array_like). In NumPy 1.20 a proper [ArrayLike type](https://numpy.org/devdocs/reference/typing.html) was defined, however. I did initially make used of `ArrayLike` typing but removed it since (1) it doesn't (yet) support hinting array shape or dtype, and (2) it requires a bump to NumPy 1.20.

I'll note the term `array_like` is still used in the NumPy docs, and is [already in use in PyVista](https://github.com/search?q=org%3Apyvista%20array_like&type=code)

Do you think we should include a link to the glossary definition for reference?

Reply:
I like having a link somewhere, especially if it's only going to get more common in the pyvista codebase

Reply:
The additional info and link has been aded in [7c8a237](https://github.com/pyvista/pyvista/pull/5100/commits/7c8a237eb46dfa18e7953ced6ee05d11ae5a86d7)